File: Engine/src/main/java/org/etieskrill/engine/graphics/assimp/Vertex.java
package org.etieskrill.engine.graphics.assimp;

import org.joml.Vector2fc;
import org.joml.Vector3fc;

import java.util.List;

public class Vertex {
    
    public static final int
            POSITION_COMPONENTS = 3,
            NORMAL_COMPONENTS = 3,
            TEXTURE_COMPONENTS = 2,
            COMPONENTS = 8;

    private final Vector3fc position;
    private final Vector3fc normal;
    private final Vector2fc textureCoords;

    public Vertex(Vector3fc position, Vector3fc normal, Vector2fc textureCoords) {
        this.position = position;
        this.normal = normal;
        this.textureCoords = textureCoords;
    }

    public Vector3fc getPosition() {
        return position;
    }

    public Vector3fc getNormal() {
        return normal;
    }

    public Vector2fc getTextureCoords() {
        return textureCoords;
    }
    
    public List<Float> toList() { //TODO optimise this
        return List.of(
                position.x(), position.y(), position.z(),
                normal.x(), normal.y(), normal.z(),
                textureCoords.x(), textureCoords.y()
        );
    }
    
    @Override
    public String toString() {
        return "pos=" + position.toString() + ", norm=" + normal.toString() + ", tex=" + textureCoords.toString();
    }
    
}

File: Engine/src/commonMain/kotlin/org/etieskrill/engine/graphics/model/Vertex.kt
package org.etieskrill.engine.graphics.model

import org.etieskrill.engine.graphics.buffer.VertexArrayAccessor
import org.joml.Vector2fc
import org.joml.Vector3fc
import org.joml.Vector4fc
import org.joml.Vector4ic
import kotlin.math.pow
import kotlin.math.roundToLong

data class Vertex(
    val position: Vector3fc,
    val normal: Vector3fc? = null,
    val textureCoords: Vector2fc? = null,
    val tangent: Vector3fc? = null,
    val biTangent: Vector3fc? = null,
    val bones: Vector4ic? = null,
    val boneWeights: Vector4fc? = null,
) {

    companion object {
        object Accessor : VertexArrayAccessor<Vertex>() {
            override fun registerFields() {
                addField<Vector3fc> { it, buffer -> buffer += it.position }
                addField<Vector3fc> { it, buffer -> buffer += it.normal }
                addField<Vector2fc> { it, buffer -> buffer += it.textureCoords }
                addField<Vector3fc> { it, buffer -> buffer += it.tangent }
                addField<Vector3fc> { it, buffer -> buffer += it.biTangent }
                addField<Vector4ic> { it, buffer -> buffer += it.bones }
                addField<Vector4fc> { it, buffer -> buffer += it.boneWeights }
            }
        }
    }

    override fun toString() =
        """Vertex{
            |position=$position, 
            |normal=$normal, 
            |textureCoords=textureCoords, 
            |bones=(${bones?.x()}, ${bones?.y()}, ${bones?.z()}, ${bones?.w()}), 
            |boneWeights=(${boneWeights?.x()?.round(3)}, ${boneWeights?.y()?.round(3)}, 
            |${boneWeights?.z()?.round(3)}, ${boneWeights?.w()?.round(3)})}
        """.trimMargin()

}

private fun Float.round(decimals: Int): Float {
    val decimalPower = 10f.pow(decimals)
    return (this * decimalPower).roundToLong() / decimalPower
}
